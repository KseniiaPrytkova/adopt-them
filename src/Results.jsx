import { useState, useEffect } from 'react';
import Pet from './Pet';

const Results = ({ pets }) => {
    const ITEMS_PER_PAGE = 8;
    const MAX_ITEMS = pets.length;

    const [pageConfig, setPageConfig] = useState({
        page: 1,
        from: 0,
        to: ITEMS_PER_PAGE
    });

    useEffect(() => {
        setPageConfig({
            page: 1,
            from: 0,
            to: ITEMS_PER_PAGE
        });
    }, [pets]);

    return (
        <section className="mx-2 grid grid-cols-1 justify-items-stretch gap-4 rounded-lg bg-light-lightNavy p-4 dark:bg-dark-lightGrey  sm:grid-cols-2 lg:col-span-8 lg:grid-cols-3 lg:grid-rows-2 xl:col-span-9 xl:col-start-4 xl:grid-cols-4">
            {!pets.length ? (
                <div className="flex items-center justify-center p-10">
                    <h1>No Pets Found</h1>
                </div>
            ) : (
                pets.slice(pageConfig.from, pageConfig.to).map((pet) => {
                    return (
                        <Pet
                            animal={pet.animal}
                            key={pet.id}
                            name={pet.name}
                            breed={pet.breed}
                            images={pet.images}
                            location={`${pet.city}, ${pet.state}`}
                            id={pet.id}
                        />
                    );
                })
            )}

            {pets.length ? (
                <div className="flex items-center justify-center gap-4 sm:col-span-2 lg:col-span-3 lg:row-start-4 xl:col-span-4">
                    <button
                        disabled={pageConfig.from === 0}
                        onClick={() => {
                            setPageConfig({
                                ...pageConfig,
                                page: pageConfig.page - 1,
                                from: pageConfig.from - ITEMS_PER_PAGE,
                                to: pageConfig.to - ITEMS_PER_PAGE
                            });
                        }}
                        className="text-light-darkNavy disabled:cursor-not-allowed dark:text-dark-darkRed"
                    >
                        Prev
                    </button>
                    <span className="flex h-10 w-10 items-center justify-center text-light-darkNavy dark:text-dark-darkRed">
                        {pageConfig.page}
                    </span>
                    <button
                        disabled={pageConfig.to >= MAX_ITEMS}
                        onClick={() => {
                            setPageConfig({
                                ...pageConfig,
                                page: pageConfig.page + 1,
                                from: pageConfig.from + ITEMS_PER_PAGE,
                                to: pageConfig.to + ITEMS_PER_PAGE
                            });
                        }}
                        className="text-light-darkNavy disabled:cursor-not-allowed dark:text-dark-darkRed"
                    >
                        Next
                    </button>
                </div>
            ) : null}
        </section>
    );
};

export default Results;
