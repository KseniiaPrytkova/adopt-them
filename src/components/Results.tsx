import { Pet as PetType } from '../APIResponsesTypes';
import Pet from './Pet';

interface ResultsProps {
    pets: PetType[];
    page: number;
    setPage: (value: number | ((old: number) => number)) => void;
    isLoading: boolean;
    isFetching: boolean;
    isPreviousData: boolean;
    isError: boolean;
    error: Error;
    resultsRef: React.RefObject<HTMLElement>;
    scrollToTop: () => void;
}

const Results = ({
    pets,
    page,
    setPage,
    isLoading,
    isFetching,
    isPreviousData,
    isError,
    error,
    resultsRef,
    scrollToTop
}: ResultsProps) => {
    const handleButtonClick = (pageUpdater: (old: number) => number) => {
        scrollToTop();
        setTimeout(() => {
            setPage((old) => pageUpdater(old));
        }, 1000);
    };

    return (
        <section
            ref={resultsRef}
            className="mx-2 flex flex-col rounded-lg bg-light-lightNavy p-2 dark:bg-dark-lightGrey lg:col-span-8 xl:col-span-9"
        >
            <article className="m-2 grid basis-52 gap-4 sm:grid-cols-2 md:grid-cols-3 lg:basis-full xl:grid-cols-4 xl:grid-rows-2">
                {isError ? (
                    <div>Error: {error.message} </div>
                ) : isFetching || isLoading ? (
                    <div>Loading...</div>
                ) : !pets.length ? (
                    <div>No Pets FoundðŸ˜©</div>
                ) : (
                    pets.map((pet, index) => {
                        return (
                            <Pet
                                animal={pet.animal}
                                key={pet.id}
                                name={pet.name}
                                breed={pet.breed}
                                images={pet.images}
                                location={`${pet.city}, ${pet.state}`}
                                id={pet.id}
                                delay={index * 100}
                                currentPage={page}
                            />
                        );
                    })
                )}
            </article>

            <nav className="mb-2 flex justify-center gap-4">
                <button
                    data-testid={`prev-button-${page}`}
                    type="button"
                    onClick={() =>
                        handleButtonClick((old) => Math.max(old - 1, 0))
                    }
                    disabled={page === 0}
                    className={`text-light-darkNavy dark:text-dark-darkRed ${
                        page === 0
                            ? 'cursor-not-allowed opacity-50'
                            : ' opacity-100'
                    }`}
                >
                    Prev
                </button>

                {isFetching ? (
                    <span data-testid="spinner" className="inline animate-spin">
                        {page + 1}
                    </span>
                ) : (
                    <span className="flex h-10 w-10 items-center justify-center text-light-darkNavy dark:text-dark-darkRed">
                        {page + 1}
                    </span>
                )}

                <button
                    type="button"
                    data-testid={`next-button-${page}`}
                    onClick={() => handleButtonClick((old) => old + 1)}
                    disabled={isFetching || isPreviousData || pets.length < 10}
                    className={`text-light-darkNavy dark:text-dark-darkRed ${
                        isFetching || isPreviousData || pets.length < 10
                            ? 'cursor-not-allowed opacity-50'
                            : 'opacity-100'
                    }`}
                >
                    Next
                </button>
            </nav>
        </section>
    );
};

export default Results;
