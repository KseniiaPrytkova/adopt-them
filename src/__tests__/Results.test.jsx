import { expect, test, vi, beforeEach, afterEach, describe } from 'vitest';
import { render, waitFor } from '@testing-library/react';
import { createRenderer } from 'react-test-renderer/shallow';
import { StaticRouter } from 'react-router-dom/server';
import Results from '../components/Results';

const executeAfterOneSecond = (func, arg) => {
    setTimeout(() => {
        func(arg);
    }, 1000 + 100);
};

describe('Results component', () => {
    const mockSetPage = vi.fn();
    const mockScrollToTop = vi.fn();

    beforeEach(() => {
        vi.useFakeTimers();
        mockSetPage.mockReset();
        mockScrollToTop.mockReset();
    });

    afterEach(() => {
        vi.restoreAllMocks();
        vi.runAllTimers();
        vi.useRealTimers();
    });

    const defaultProps = {
        pets: [],
        page: 0,
        setPage: mockSetPage,
        isLoading: false,
        isFetching: false,
        isPreviousData: false,
        isError: false,
        error: null,
        resultsRef: null,
        scrollToTop: mockScrollToTop
    };

    const pets = [
        {
            id: 1,
            name: 'Luna',
            animal: 'dog',
            city: 'Seattle',
            state: 'WA',
            description:
                "Luna is actually the most adorable dog in the world. Her hobbies include yelling at squirrels, aggressively napping on her owners' laps, and asking to be fed two hours before IT'S DAMN WELL TIME LUNA. Luna is beloved by her puppy parents and lazily resides currently in Seattle, Washington.",
            breed: 'Havanese',
            images: [
                'http://pets-images.dev-apis.com/pets/dog25.jpg',
                'http://pets-images.dev-apis.com/pets/dog26.jpg',
                'http://pets-images.dev-apis.com/pets/dog27.jpg',
                'http://pets-images.dev-apis.com/pets/dog28.jpg',
                'http://pets-images.dev-apis.com/pets/dog29.jpg'
            ]
        },
        {
            id: 2,
            name: 'Charisse',
            animal: 'rabbit',
            city: 'Lexington',
            state: 'KY',
            description:
                'Maecenas leo odio, condimentum id, luctus nec, molestie sed, justo. Pellentesque viverra pede ac diam. Cras pellentesque volutpat dui.\n\nMaecenas tristique, est et tempus semper, est quam pharetra magna, ac consequat metus sapien ut nunc. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Mauris viverra diam vitae quam. Suspendisse potenti.',
            breed: 'Havanese',
            images: ['http://pets-images.dev-apis.com/pets/rabbit0.jpg']
        },
        {
            id: 3,
            name: 'Maitilde',
            animal: 'rabbit',
            city: 'Dallas',
            state: 'TX',
            description:
                'Duis consequat dui nec nisi volutpat eleifend. Donec ut dolor. Morbi vel lectus in quam fringilla rhoncus.\n\nMauris enim leo, rhoncus sed, vestibulum sit amet, cursus id, turpis. Integer aliquet, massa id lobortis convallis, tortor risus dapibus augue, vel accumsan tellus nisi eu orci. Mauris lacinia sapien quis libero.\n\nNullam sit amet turpis elementum ligula vehicula consequat. Morbi a ipsum. Integer a nibh.',
            breed: 'Lab',
            images: ['http://pets-images.dev-apis.com/pets/rabbit1.jpg']
        },
        {
            id: 4,
            name: 'Natalina',
            animal: 'rabbit',
            city: 'Tampa',
            state: 'FL',
            description:
                'Mauris enim leo, rhoncus sed, vestibulum sit amet, cursus id, turpis. Integer aliquet, massa id lobortis convallis, tortor risus dapibus augue, vel accumsan tellus nisi eu orci. Mauris lacinia sapien quis libero.\n\nNullam sit amet turpis elementum ligula vehicula consequat. Morbi a ipsum. Integer a nibh.\n\nIn quis justo. Maecenas rhoncus aliquam lacus. Morbi quis tortor id nulla ultrices aliquet.',
            breed: 'Lab',
            images: ['http://pets-images.dev-apis.com/pets/rabbit2.jpg']
        },
        {
            id: 5,
            name: 'Michail',
            animal: 'reptile',
            city: 'Tuscaloosa',
            state: 'AL',
            description:
                'Nulla ut erat id mauris vulputate elementum. Nullam varius. Nulla facilisi.\n\nCras non velit nec nisi vulputate nonummy. Maecenas tincidunt lacus at velit. Vivamus vel nulla eget eros elementum pellentesque.\n\nQuisque porta volutpat erat. Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla. Nunc purus.',
            breed: 'Havanese',
            images: ['http://pets-images.dev-apis.com/pets/reptile1.jpg']
        },
        {
            id: 6,
            name: 'Gizela',
            animal: 'bird',
            city: 'Carol Stream',
            state: 'IL',
            description:
                'Nullam sit amet turpis elementum ligula vehicula consequat. Morbi a ipsum. Integer a nibh.\n\nIn quis justo. Maecenas rhoncus aliquam lacus. Morbi quis tortor id nulla ultrices aliquet.',
            breed: 'Havanese',
            images: ['http://pets-images.dev-apis.com/pets/bird2.jpg']
        },
        {
            id: 7,
            name: 'Laughton',
            animal: 'reptile',
            city: 'Bridgeport',
            state: 'CT',
            description:
                'Cras mi pede, malesuada in, imperdiet et, commodo vulputate, justo. In blandit ultrices enim. Lorem ipsum dolor sit amet, consectetuer adipiscing elit.',
            breed: 'Havanese',
            images: ['http://pets-images.dev-apis.com/pets/reptile2.jpg']
        },
        {
            id: 8,
            name: 'Si',
            animal: 'dog',
            city: 'Charlotte',
            state: 'NC',
            description:
                'Praesent id massa id nisl venenatis lacinia. Aenean sit amet justo. Morbi ut odio.',
            breed: 'Lab',
            images: ['http://pets-images.dev-apis.com/pets/dog0.jpg']
        },
        {
            id: 9,
            name: 'Lyda',
            animal: 'rabbit',
            city: 'Springfield',
            state: 'IL',
            description:
                'Curabitur gravida nisi at nibh. In hac habitasse platea dictumst. Aliquam augue quam, sollicitudin vitae, consectetuer eget, rutrum at, lorem.\n\nInteger tincidunt ante vel ipsum. Praesent blandit lacinia erat. Vestibulum sed magna at nunc commodo placerat.',
            breed: 'Lab',
            images: ['http://pets-images.dev-apis.com/pets/rabbit3.jpg']
        },
        {
            id: 10,
            name: 'Jackquelin',
            animal: 'dog',
            city: 'Tucson',
            state: 'AZ',
            description:
                'Donec diam neque, vestibulum eget, vulputate ut, ultrices vel, augue. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Donec pharetra, magna vestibulum aliquet ultrices, erat tortor sollicitudin mi, sit amet lobortis sapien sapien non mi. Integer ac neque.',
            breed: 'Lab',
            images: ['http://pets-images.dev-apis.com/pets/dog1.jpg']
        }
    ];

    test('renders correctly with no pets', () => {
        const { asFragment } = render(<Results pets={[]} page={0} />);
        expect(asFragment()).toMatchSnapshot();
    });

    // test('renders correctly with some pets', () => {
    //     const { asFragment } = render(
    //         <StaticRouter>
    //             <Results pets={pets} />
    //         </StaticRouter>
    //     );
    //     expect(asFragment()).toMatchSnapshot();
    // });

    // with shallow rendering
    test('renders correctly with some pets', () => {
        const r = createRenderer();
        r.render(<Results pets={pets} />);
        expect(r.getRenderOutput()).toMatchSnapshot();
    });

    test('pets appear smoothly in the correct order', async () => {
        const results = render(
            <StaticRouter>
                <Results {...defaultProps} pets={pets} />
            </StaticRouter>
        );

        pets.forEach((pet, index) => {
            const petThumbnail = results.getByTestId(`thumbnail-${pet.id}`);
            const expectedDelay = index * 100;

            expect(petThumbnail.closest('a').style.animationDelay).toBe(
                `${expectedDelay}ms`
            );
        });

        results.unmount();
    });

    test('renders error message if there is an error', () => {
        const results = render(
            <StaticRouter>
                <Results
                    {...defaultProps}
                    isError={true}
                    error={{ message: 'this is an error' }}
                />
            </StaticRouter>
        );

        // console.log(results.container.innerHTML);
        expect(results.getByText('Error: this is an error')).toBeTruthy();

        results.unmount();
    });

    test('renders loading message', () => {
        const results = render(
            <StaticRouter>
                <Results {...defaultProps} isLoading={true} />
            </StaticRouter>
        );

        expect(results.getByText('Loading...')).toBeTruthy();

        results.unmount();
    });

    test('handleButtonClick for previous page', async () => {
        const results = render(
            <StaticRouter>
                <Results
                    {...defaultProps}
                    scrollToTop={mockScrollToTop}
                    setPage={mockSetPage}
                    pets={pets}
                    page={1}
                />
            </StaticRouter>
        );

        const prevButton = await results.findByTestId('prev-button-1');
        prevButton.click();
        expect(mockScrollToTop).toHaveBeenCalled();

        executeAfterOneSecond(mockSetPage, 0);
        vi.advanceTimersByTime(1100);
        await waitFor(() => expect(mockSetPage).toHaveBeenCalled());
        const setPageCallback = mockSetPage.mock.calls[0][0];
        expect(setPageCallback(1)).toEqual(0);

        results.unmount();
    });

    test('handleButtonClick for next page', async () => {
        const results = render(
            <StaticRouter>
                <Results
                    {...defaultProps}
                    scrollToTop={mockScrollToTop}
                    setPage={mockSetPage}
                    pets={pets}
                    page={1}
                />
            </StaticRouter>
        );

        const nextButton = await results.findByTestId('next-button-1');
        nextButton.click();
        expect(mockScrollToTop).toHaveBeenCalled();

        executeAfterOneSecond(mockSetPage, 2);
        vi.advanceTimersByTime(1100);
        await waitFor(() => expect(mockSetPage).toHaveBeenCalled());
        const setPageCallback = mockSetPage.mock.calls[0][0];
        expect(setPageCallback(1)).toEqual(2);

        results.unmount();
    });

    test('displays spinner when isFetching is true', () => {
        const { getByTestId } = render(
            <StaticRouter>
                <Results {...defaultProps} isFetching={true} page={1} />
            </StaticRouter>
        );

        const spinner = getByTestId('spinner');

        expect(spinner).toBeTruthy();
        expect(Array.from(spinner.classList)).toContain(
            'inline',
            'animate-spin'
        );
        expect(spinner.textContent).toBe('2');
    });
});
