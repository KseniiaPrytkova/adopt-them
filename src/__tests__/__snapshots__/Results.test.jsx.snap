// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly with no pets 1`] = `
<DocumentFragment>
  <section
    class="mx-2 flex flex-col rounded-lg bg-light-lightNavy p-2 dark:bg-dark-lightGrey lg:col-span-8 xl:col-span-9"
  >
    <article
      class="m-2 grid basis-52 gap-4 sm:grid-cols-2 md:grid-cols-3 lg:basis-full xl:grid-cols-4 xl:grid-rows-2"
    >
      <div>
        No Pets FoundðŸ˜©
      </div>
    </article>
    <nav
      class="mb-2 flex justify-center gap-4"
    >
      <button
        class="text-light-darkNavy dark:text-dark-darkRed  opacity-100"
        type="button"
      >
        Prev
      </button>
      <span
        class="flex h-10 w-10 items-center justify-center text-light-darkNavy dark:text-dark-darkRed"
      >
        NaN
      </span>
      <button
        class="text-light-darkNavy dark:text-dark-darkRed cursor-not-allowed opacity-50"
        disabled=""
        type="button"
      >
        Next
      </button>
    </nav>
  </section>
</DocumentFragment>
`;

exports[`renders correctly with some pets 1`] = `
<section
  className="mx-2 flex flex-col rounded-lg bg-light-lightNavy p-2 dark:bg-dark-lightGrey lg:col-span-8 xl:col-span-9"
>
  <article
    className="m-2 grid basis-52 gap-4 sm:grid-cols-2 md:grid-cols-3 lg:basis-full xl:grid-cols-4 xl:grid-rows-2"
  >
    <Pet
      animal="dog"
      breed="Havanese"
      delay={0}
      id={1}
      images={
        [
          "http://pets-images.dev-apis.com/pets/dog25.jpg",
          "http://pets-images.dev-apis.com/pets/dog26.jpg",
          "http://pets-images.dev-apis.com/pets/dog27.jpg",
          "http://pets-images.dev-apis.com/pets/dog28.jpg",
          "http://pets-images.dev-apis.com/pets/dog29.jpg",
        ]
      }
      location="Seattle, WA"
      name="Luna"
    />
    <Pet
      animal="rabbit"
      breed="Havanese"
      delay={100}
      id={2}
      images={
        [
          "http://pets-images.dev-apis.com/pets/rabbit0.jpg",
        ]
      }
      location="Lexington, KY"
      name="Charisse"
    />
    <Pet
      animal="rabbit"
      breed="Lab"
      delay={200}
      id={3}
      images={
        [
          "http://pets-images.dev-apis.com/pets/rabbit1.jpg",
        ]
      }
      location="Dallas, TX"
      name="Maitilde"
    />
    <Pet
      animal="rabbit"
      breed="Lab"
      delay={300}
      id={4}
      images={
        [
          "http://pets-images.dev-apis.com/pets/rabbit2.jpg",
        ]
      }
      location="Tampa, FL"
      name="Natalina"
    />
    <Pet
      animal="reptile"
      breed="Havanese"
      delay={400}
      id={5}
      images={
        [
          "http://pets-images.dev-apis.com/pets/reptile1.jpg",
        ]
      }
      location="Tuscaloosa, AL"
      name="Michail"
    />
    <Pet
      animal="bird"
      breed="Havanese"
      delay={500}
      id={6}
      images={
        [
          "http://pets-images.dev-apis.com/pets/bird2.jpg",
        ]
      }
      location="Carol Stream, IL"
      name="Gizela"
    />
    <Pet
      animal="reptile"
      breed="Havanese"
      delay={600}
      id={7}
      images={
        [
          "http://pets-images.dev-apis.com/pets/reptile2.jpg",
        ]
      }
      location="Bridgeport, CT"
      name="Laughton"
    />
    <Pet
      animal="dog"
      breed="Lab"
      delay={700}
      id={8}
      images={
        [
          "http://pets-images.dev-apis.com/pets/dog0.jpg",
        ]
      }
      location="Charlotte, NC"
      name="Si"
    />
    <Pet
      animal="rabbit"
      breed="Lab"
      delay={800}
      id={9}
      images={
        [
          "http://pets-images.dev-apis.com/pets/rabbit3.jpg",
        ]
      }
      location="Springfield, IL"
      name="Lyda"
    />
    <Pet
      animal="dog"
      breed="Lab"
      delay={900}
      id={10}
      images={
        [
          "http://pets-images.dev-apis.com/pets/dog1.jpg",
        ]
      }
      location="Tucson, AZ"
      name="Jackquelin"
    />
  </article>
  <nav
    className="mb-2 flex justify-center gap-4"
  >
    <button
      className="text-light-darkNavy dark:text-dark-darkRed  opacity-100"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Prev
    </button>
    <span
      className="flex h-10 w-10 items-center justify-center text-light-darkNavy dark:text-dark-darkRed"
    >
      NaN
    </span>
    <button
      className="text-light-darkNavy dark:text-dark-darkRed opacity-100"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Next
    </button>
  </nav>
</section>
`;
